cosmos/operators/watcher.py

```diff
@@ -116,3 +124,151 @@ def _callback(ev: EventMsg) -> None:
+        super().__init__(
+            poke_interval=poke_interval,
+            timeout=timeout,
+            profile_config=profile_config,
+            project_dir=project_dir,
+            profiles_dir=profiles_dir,
+            **kwargs,
+        )
+        self.model_unique_id = extra_context.get("dbt_node_config", {}).get("unique_id")
+        self.producer_task_id = producer_task_id
+
+    @staticmethod
+    def _filter_flags(flags: list[str]) -> list[str]:
+        """Filters out dbt flags that are incompatible with retry (e.g., --select, --exclude)."""
+        filtered = []
+        skip_next = False
+        for token in flags:
+            if skip_next:
+                if token.startswith("--"):
+                    skip_next = False
+                else:
+                    continue  # skip value of previous flag
+            if token in ("--select", "--exclude"):
+                skip_next = True
+                continue
+            filtered.append(token)
+        return filtered
+
+    def _handle_task_retry(self, try_number: int, context: Context) -> bool:
+        """
+        Handles logic for retrying a failed dbt model execution.
+        Reconstructs the dbt command by cloning the project and re-running the model
+        with appropriate flags, while ensuring flags like `--select` or `--exclude` are excluded.
+        """
+        self.log.info(
+            "Retry attempt #%s – Re-running model '%s' from project '%s'",
+            try_number - 1,
+            self.model_unique_id,
+            self.project_dir,
+        )
+
+        upstream_task = context["ti"].task.dag.get_task(self.producer_task_id)
+
+        extra_flags: list[str] = []
+        if upstream_task and hasattr(upstream_task, "add_cmd_flags"):
+            raw_flags = upstream_task.add_cmd_flags()
+            extra_flags = self._filter_flags(raw_flags)
+
+        model_selector = self.model_unique_id.split(".")[-1]
+        cmd_flags = extra_flags + ["--select", model_selector]
+
+        self.build_and_run_cmd(context, cmd_flags=cmd_flags)
+
+        self.log.info("dbt run completed successfully on retry for model '%s'", self.model_unique_id)
+        return True
+
+    def _get_status_from_events(self, ti: Any) -> Any:
+
+        dbt_startup_events = ti.xcom_pull(task_ids=self.producer_task_id, key="dbt_startup_events")
+        if dbt_startup_events:  # pragma: no cover
+            self.log.info("Dbt Startup Event: %s", dbt_startup_events)
+
+        node_finished_key = f"nodefinished_{self.model_unique_id.replace('.', '__')}"
+        compressed_b64_event_msg = ti.xcom_pull(task_ids=self.producer_task_id, key=node_finished_key)
+
+        if not compressed_b64_event_msg:
+            return None
+
+        compressed_bytes = base64.b64decode(compressed_b64_event_msg)
+        event_json_str = zlib.decompress(compressed_bytes).decode("utf-8")
+        event_json = ast.literal_eval(event_json_str)
```

Comment:
@pankajastro why we are using `ast.literal_eval` instead of something like `json.loads`?

Reply:
vibe coding!! yes json should work